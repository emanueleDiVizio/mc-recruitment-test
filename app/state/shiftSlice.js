import { createSlice } from '@reduxjs/toolkit'

const shiftSlice = createSlice({
  name: 'shift',
  initialState: { onShift: false },
  reducers: {
    startShift: state => ({
      ...state,
      onShift: true,
    }),
    endShift: state => ({
      ...state,
      onShift: false,
    }),
  },
})

export default shiftSlice
