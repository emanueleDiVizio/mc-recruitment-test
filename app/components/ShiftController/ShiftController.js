import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { View, Text, Button } from 'react-native'
import shiftSlice from '../../state/shiftSlice'

const shiftStatusSelector = state => state.shift.onShift

const ShiftController = () => {
  const dispatch = useDispatch()
  const onUserPressButton = () => {
    dispatch(shiftSlice.actions.startShift())
  }
  const shiftStatus = useSelector(shiftStatusSelector)
  return (
    <View>
      <Text>
        Please press the button below to start your 8 hours shift, you will
        receive a notification to wash your hand every 2 hours. There is
        currently no way to end the shift prematurely.
      </Text>
      <Button
        onPress={onUserPressButton}
        testID="shift-button"
        title="Start shift"
        disabled={shiftStatus}
      />
      <Text>Shift status: {shiftStatus ? 'ON' : 'OFF'}</Text>
    </View>
  )
}

export default ShiftController
