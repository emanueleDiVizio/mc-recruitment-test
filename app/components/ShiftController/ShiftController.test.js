import React from 'react'
import ShiftController from './ShiftController'
import { fireEvent, render } from '@testing-library/react-native'
import * as redux from 'react-redux'
import shiftSlice from '../../state/shiftSlice'

describe('HandWashComponent', () => {
  let state
  let dispatch

  beforeEach(() => {
    dispatch = jest.fn()
    redux.useSelector = selector => selector(state)
    redux.useDispatch = () => dispatch
  })

  it('renders correctly when off shift', () => {
    state = {
      shift: { onShift: false },
    }
    const { baseElement } = render(<ShiftController />)

    expect(baseElement).toMatchSnapshot()
  })

  it('renders correctly when on shift', () => {
    state = {
      shift: { onShift: true },
    }
    const { baseElement } = render(<ShiftController />)

    expect(baseElement).toMatchSnapshot()
  })

  it('dispatches `startShift` action when user presses button', async () => {
    state = {
      shift: { onShift: false },
    }
    const { baseElement, getByTestId } = render(<ShiftController />)

    const button = getByTestId('shift-button')
    fireEvent.press(button)

    expect(dispatch).toHaveBeenCalledWith(shiftSlice.actions.startShift())

    expect(baseElement).toMatchSnapshot()
  })
})
