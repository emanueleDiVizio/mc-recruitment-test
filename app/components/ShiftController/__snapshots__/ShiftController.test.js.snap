// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HandWashComponent dispatches \`startShift\` action when user presses button 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <Text>
        Please press the button below to start your 8 hours shift, you will receive a notification to wash your hand every 2 hours
      </Text>
      <Button
        testID="shift-button"
        title="Start shift"
      />
      <Text>
        Shift status: 
        OFF
      </Text>
    </View>
  </View>
</View>
`;

exports[`HandWashComponent renders correctly when off shift 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <Text>
        Please press the button below to start your 8 hours shift, you will receive a notification to wash your hand every 2 hours
      </Text>
      <Button
        testID="shift-button"
        title="Start shift"
      />
      <Text>
        Shift status: 
        OFF
      </Text>
    </View>
  </View>
</View>
`;

exports[`HandWashComponent renders correctly when on shift 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <Text>
        Please press the button below to start your 8 hours shift, you will receive a notification to wash your hand every 2 hours
      </Text>
      <Button
        testID="shift-button"
        title="Start shift"
      />
      <Text>
        Shift status: 
        ON
      </Text>
    </View>
  </View>
</View>
`;
